src/main/java/org/elasticsearch/indices/SyncedFlushService.java

```diff
@@ -0,0 +1,755 @@
+        if (primaryNode == null) {
+            logger.trace("{} failed to resolve node for primary shard {}, skipping sync", shardId, primaryShard);
+            return -1;
+        }
+        final AtomicInteger result = new AtomicInteger(-1);
+        final CountDownLatch latch = new CountDownLatch(1);
+        logger.trace("{} retrieving in flight operation count", shardId);
+        transportService.sendRequest(primaryNode, IN_FLIGHT_OPS_ACTION_NAME, new InFlightOpsRequest(shardId),
+                new BaseTransportResponseHandler<InFlightOpsResponse>() {
+                    @Override
+                    public InFlightOpsResponse newInstance() {
+                        return new InFlightOpsResponse();
+                    }
+
+                    @Override
+                    public void handleResponse(InFlightOpsResponse response) {
+                        result.set(response.opCount());
+                        latch.countDown();
+                    }
+
+                    @Override
+                    public void handleException(TransportException exp) {
+                        logger.debug("{} unexpected error while retrieving inflight op count", shardId);
+                        result.set(-1);
+                        latch.countDown();
+                    }
+
+                    @Override
+                    public String executor() {
+                        return ThreadPool.Names.SAME;
+                    }
+                });
+        try {
+            if (latch.await(inflightOpsTimeout.millis(), TimeUnit.MILLISECONDS) == false) {
+                logger.debug("{} in flight operation check timed out after [{}]", shardId, syncTimeout);
+            }
+        } catch (InterruptedException e) {
+            logger.debug("{} interrupted while waiting for in flight operation check", shardId);
+        }
+        final int count = result.get();
+        logger.trace("{} in flight operation count [{}]", shardId, count);
+        return count;
+    }
+
+
+    void sendSyncRequests(final String syncId, List<ShardRouting> shards, ClusterState state, Map<String, byte[]> expectedCommitIds, final ShardId shardId, final ActionListener<SyncedFlushResult> listener) {
+        final CountDown countDownLatch = new CountDown(shards.size());
+        final Map<ShardRouting, SyncedFlushResponse> results = ConcurrentCollections.newConcurrentMap();
+        for (final ShardRouting shard : shards) {
+            final DiscoveryNode node = state.nodes().get(shard.currentNodeId());
+            if (node == null) {
+                logger.trace("{} is assigned to an unknown node. skipping for sync id [{}]. shard routing {}", shardId, syncId, shard);
+                results.put(shard, new SyncedFlushResponse("unknown node"));
+                if (countDownLatch.countDown()) {
+                    listener.onResponse(new SyncedFlushResult(shardId, syncId, results));
+                }
+                continue;
```

Comment:
if we call the listener we should be done here right, can we assert this somehow? ie `assert shard.equals(shards.get(shards.size()-1);`